### ISSUE TITLE
BUG: stack over flow when getAllMethods()

### ISSUE THREAD

**Issue by danglotb** [INRIA/spoon#2378]
Hi all, I am reporting an error:

Spoon throws a `StackOverFlowError` when I am trying to call `getAllMethods()` on the following object:

```java
public class Matriochka extends SubMatriochka {
     public void aMethod()  {
     }
 }
```
```java
public class SubMatriochka {
     public <F extends Enum<F> & Factory> int aMethod(Class<F> object) {
        return 23;
    }
 }
```

_i.e._ the following snippet is throwing a  `StackOverFlowError`
```java
final Class<?> aClass = Matriochka.class;
final CtType<?> ctType = createFactory().Class().get(aClass);
ctType.getAllMethods();
```

You may find a test, with all the resources to reproduce the error on my [fork](https://github.com/danglotb/spoon/tree/stackoverflow-error) on the branch **stackoverflow-error**.

Thank you!

**Comment by pvojtechovsky** [comment]
I guess it is related to known problem discussed in #2161. The solution needs some changes in Spoon model - so it needs some time.

**Comment by monperrus** [comment]
fixed in #2406. don't hesitate to reopen if it does not work the the latest master.
